### PR TITLE
Add license feature gating

### PR DESCRIPTION
**Pull request by danvixent**

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -545,6 +543,8 @@ func (a *ApplicationHandler) BuildDataPlaneRoutes() *chi.Mux {
 		uiRouter.Use(middleware.JsonResponse)
 		uiRouter.Use(chiMiddleware.Maybe(middleware.RequireAuth(), shouldAuthRoute))
 
+		uiRouter.Get("/license_features", handler.GetLicenseFeatures)
```
Comment (jirevwe) [inline api/api.go:546]: this is a nitpick, but shouldn't this be
```suggestion
		uiRouter.Get("/licenses/{id}/features", handler.GetLicenseFeatures)
		
		or 
		
		uiRouter.Get("/licenses/features", handler.GetLicenseFeatures)
```

File: api/handlers/license.go [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+import (
+	"net/http"
+
+	"github.com/frain-dev/convoy/util"
+	"github.com/go-chi/render"
+)
+
+func (h *Handler) GetLicenseFeatures(w http.ResponseWriter, r *http.Request) {
+	_ = render.Render(w, r, util.NewServerResponse("Login successful", h.A.Licenser.FeatureListJSON(), http.StatusOK))
```
[the remark this answers is not in the document]
Comment (jirevwe) [reply api/handlers/license.go:11]: +1

File: api/handlers/shim.go
Comment (jirevwe) [inline api/handlers/shim.go:1]: Will the removal break anything?

File: internal/pkg/license/keygen/feature.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	SynchronousWebhooks     Feature = "SYNCHRONOUS_WEBHOOKS"
+)
+
+const (
+	CommunityPlan  PlanType = "community"
+	BusinessPlan   PlanType = "business"
+	EnterprisePlan PlanType = "enterprise"
+)
+
+func (p PlanType) IsValid() bool {
```
Comment (jirevwe) [inline internal/pkg/license/keygen/feature.go:31]: That is a weird name. Does this mean the community plan isn't valid?

File: internal/pkg/license/keygen/keygen.go [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+)
+
+func getFeatureList(ctx context.Context, l *keygen.License) (map[Feature]Properties, error) {
+	entitlements, err := l.Entitlements(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load license entitlements: %v", err)
+	}
+
+	if len(entitlements) == 0 {
+		return nil, fmt.Errorf("license has no entitlements")
```
Comment (jirevwe) [inline internal/pkg/license/keygen/keygen.go:163]: Should this be an error case?
Can't a license have zero entitlements?

File: internal/pkg/license/noop/noop.go
Comment (jirevwe) [inline internal/pkg/license/noop/noop.go:1]: Can we add the integration build tag to this so it only compiles when we atr testing

**Review by jirevwe — CHANGES_REQUESTED** [review]
I left a few comments; it looks good overall, though. Good job

File: internal/pkg/license/keygen/keygen.go [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+}
+
+type Config struct {
+	LicenseKey    string
+	OrgRepo       datastore.OrganisationRepository
+	OrgMemberRepo datastore.OrganisationMemberRepository
+}
+
+func init() {
+	keygen.Account = "cb3ba418-a9a0-4aa1-8c20-a844a06eeac9"
```
[the remark this answers is not in the document]
Comment (jirevwe) [reply internal/pkg/license/keygen/keygen.go:45]: Yeah, they are meant to be checked in

**Review by jirevwe — APPROVED** [review]

File: api/handlers/license.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (danvixent) [reply api/handlers/license.go]: fixed

File: api/handlers/subscription.go [hunk 6]
```diff
@@ -400,6 +402,11 @@ func (h *Handler) ToggleSubscriptionStatus(w http.ResponseWriter, r *http.Reques
 //	@Security		ApiKeyAuth
 //	@Router			/v1/projects/{projectID}/subscriptions/test_filter [post]
 func (h *Handler) TestSubscriptionFilter(w http.ResponseWriter, r *http.Request) {
+	if !h.A.Licenser.AdvancedSubscriptions() {
+		_ = render.Render(w, r, util.NewErrorResponse("your instance does not have access to subscription filters, upgrade to access this feature", http.StatusBadRequest))
```
[the remark this answers is not in the document]
Comment (danvixent) [reply api/handlers/subscription.go:406]: I'm unable to find the duplicate

File: net/dispatcher.go [hunk 7]
```diff
@@ -21,7 +23,7 @@ type Dispatcher struct {
 	client *http.Client
 }
 
-func NewDispatcher(httpProxy string, enforceSecure bool) (*Dispatcher, error) {
+func NewDispatcher(httpProxy string, licenser license.Licenser, enforceSecure bool) (*Dispatcher, error) {
```
[the remark this answers is not in the document]
Comment (danvixent) [reply net/dispatcher.go:26]: I think it's better this way for mock testing.

[the remark this answers is not in the document]
Comment (danvixent) [reply net/dispatcher.go:26]: @mekilis maybe but I reckon it will be more complex to do so

**Review by jirevwe — APPROVED** [review]

**Pull request merged**